ext/dom/html_document.c

```diff
@@ -0,0 +1,1047 @@
+		case LXB_HTML_RULES_ERROR_DOTOINHENOMO: return "doctype-token-in-head-noscript-mode";
+		case LXB_HTML_RULES_ERROR_DOTOAFHEMO: return "doctype-token-after-head-mode";
+		case LXB_HTML_RULES_ERROR_HETOAFHEMO: return "head-token-after-head-mode";
+		case LXB_HTML_RULES_ERROR_DOTOINBOMO: return "doctype-token-in-body-mode";
+		case LXB_HTML_RULES_ERROR_BAENOPELISWR: return "bad-ending-open-elements-is-wrong";
+		case LXB_HTML_RULES_ERROR_OPELISWR: return "open-elements-is-wrong";
+		case LXB_HTML_RULES_ERROR_UNELINOPELST: return "unexpected-element-in-open-elements-stack";
+		case LXB_HTML_RULES_ERROR_MIELINOPELST: return "missing-element-in-open-elements-stack";
+		case LXB_HTML_RULES_ERROR_NOBOELINSC: return "no-body-element-in-scope";
+		case LXB_HTML_RULES_ERROR_MIELINSC: return "missing-element-in-scope";
+		case LXB_HTML_RULES_ERROR_UNELINSC: return "unexpected-element-in-scope";
+		case LXB_HTML_RULES_ERROR_UNELINACFOST: return "unexpected-element-in-active-formatting-stack";
+		case LXB_HTML_RULES_ERROR_UNENOFFI: return "unexpected-end-of-file";
+		case LXB_HTML_RULES_ERROR_CHINTATE: return "characters-in-table-text";
+		case LXB_HTML_RULES_ERROR_DOTOINTAMO: return "doctype-token-in-table-mode";
+		case LXB_HTML_RULES_ERROR_DOTOINSEMO: return "doctype-token-in-select-mode";
+		case LXB_HTML_RULES_ERROR_DOTOAFBOMO: return "doctype-token-after-body-mode";
+		case LXB_HTML_RULES_ERROR_DOTOINFRMO: return "doctype-token-in-frameset-mode";
+		case LXB_HTML_RULES_ERROR_DOTOAFFRMO: return "doctype-token-after-frameset-mode";
+		case LXB_HTML_RULES_ERROR_DOTOFOCOMO: return "doctype-token-foreign-content-mode";
+		default: return "unknown error";
+	}
+}
+
+static const char *dom_lexbor_libxml2_bridge_status_code_to_string(lexbor_libxml2_bridge_status status)
+{
+	switch (status) {
+		case LEXBOR_LIBXML2_BRIDGE_STATUS_CANNOT_INIT: return "cannot initialize data structures";
+		case LEXBOR_LIBXML2_BRIDGE_STATUS_FATAL_PARSE: return "fatal error in parsing";
+		case LEXBOR_LIBXML2_BRIDGE_STATUS_OVERFLOW: return "string length overflow";
+		case LEXBOR_LIBXML2_BRIDGE_STATUS_OOM: return "out of memory";
+		default: return "unknown error";
+	}
+}
+
+static void dom_reset_line_column_cache(dom_line_column_cache *cache)
+{
+	cache->last_line = 1;
+	cache->last_column = 1;
+	cache->last_offset = 0;
+}
+
+static void dom_find_line_and_column_using_cache(const dom_lexbor_libxml2_bridge_application_data *application_data, dom_line_column_cache *cache, size_t offset)
+{
+	offset -= application_data->current_total_offset;
+	if (offset > application_data->current_input_length) {
+		/* Possible with empty input, also just good for general safety */
+		offset = application_data->current_input_length;
+	}
+
+	/* Either unicode or UTF-8 data */
+	if (application_data->current_input_codepoints != NULL) {
+		while (cache->last_offset < offset) {
+			if (application_data->current_input_codepoints[cache->last_offset] == 0x000A) {
```

Comment:
Explanation of the magic value?

Reply:
FWIW: `\x0a` is LF. Even if one does not know that, I find this reasonably obvious based on the next line.